src/contrib/mlir/compiler/pass/mlir_subgraph_extraction.cpp

```diff
@@ -136,89 +138,159 @@ bool MLIRSubgraphExtractionPass::run_on_function(std::shared_ptr<Function> func)
+            auto user = input.get_node()->shared_from_this();
+            node_to_size_map[user]--;
+            if (node_to_size_map[user] == 0)
+            {
+                nodes_ready.push_back(user);
+            }
         }
-        if (TI(Parameter) == TI(*op) || TI(Result) == TI(*op))
+    }
+}
+
+void MLIRSubgraphExtractionPass::build_subgraphs(std::shared_ptr<Function> func)
+{
+    NGRAPH_DEBUG << "[CK Extract] Construct sub-graphs";
+    int current_subgraph_id = 0;
+
+    std::unordered_map<std::shared_ptr<Node>, size_t> node_to_size_map;
+    std::list<std::shared_ptr<Node>> nodes_ready;
+    bool last_op_is_supported = false;
+
+    for (auto op : func->get_ops())
+    {
+        size_t arg_count = op->get_input_size();
+        node_to_size_map[op] = arg_count;
+        if (arg_count == 0)
         {
-            continue;
+            nodes_ready.push_back(op);
         }
+    }
 
-        NGRAPH_DEBUG << "[CK Extract] Processing " << *op;
-        // supported op
-        for (auto pred : op->get_arguments())
+    bool change_mode = false;
+    while (!nodes_ready.empty())
+    {
+        for (auto it = nodes_ready.begin(); it != nodes_ready.end(); it++)
         {
-            int pred_subgraph_id = get_subgraph_id(pred);
-            if (pred_subgraph_id == -1)
+            auto node = *it;
+            if (TI(Result) == TI(*node))
             {
-                // predecessor doesn't belong to any sub-graph, it is an input
-                inputs.push_back(pred);
+                nodes_ready.erase(it);
+                break;
+            }
+            else if (TI(Parameter) == TI(*node))
+            {
+                process_successors(node, node_to_size_map, nodes_ready);
+                nodes_ready.erase(it);
+                break;
+            }
+            else if (is_supported_mlir_op(node))
+            {
+                if (last_op_is_supported)
+                {
+                    NodeVector inputs;
+                    for (auto pred : node->get_arguments())
+                    {
+                        int pred_subgraph_id = get_subgraph_id(pred);
+                        if (pred_subgraph_id != current_subgraph_id)
+                        {
+                            // predecessor doesn't belong to current sub-graph, it is an
+                            // input
+                            inputs.push_back(pred);
+                        }
+                    }
+                    // add inputs and op to current sub-graph
+                    MLIRSubgraph& current_subgraph = get_subgraph(current_subgraph_id);
+                    current_subgraph.add_node(node);
+                    current_subgraph.add_inputs(inputs);
+                    NGRAPH_DEBUG << "[CK Extract] Node Processed " << *node;
+                    process_successors(node, node_to_size_map, nodes_ready);
+                    nodes_ready.erase(it);
+                    change_mode = false;
+                    break;
```

Comment:
Do we need to break here ? Can't we just keep looking for the next supported node in the list ?
The idea is that if last node is supported (mode 0), we keep scanning the worklist and adding supported nodes to the sub-graph until we cannot find any (then it is a change of mode to mode 1). If last node is unsupported (mode 1), we keep scanning until and visiting unsupported nodes until none is found (then switch back to mode 0).

Reply:
I got segfault when tried to do that because we're removing node from the list while doing the iteration. I could try again to see if that would work.